### PR TITLE
Fixes FA, burst delays

### PR DESCRIPTION
**Pull request by Humonitarian**
## About The Pull Request

Fullauto now uses a two-layer while() system with sleep() instead of spawn()

Currently any RPM above 600 causes some bullets to be fired in pairs.
A line of code was left in we can swap to that raises the number of checks to match up with FPS rather than tickrate, which means you need an RPM of 2400 to cause bullets to pair up, but as there is no mechanical difference, I didn't see a need to.

"rapid" firerates have been changed to match the current code-locked minimum. (no ingame effect)

## Why It's Good For The Game

Firerate almost fixed (#7878 contains remaining fixes). Also optimization.

## Changelog
:cl:
tweak: firerates lower than the old minimum set to minimum to avoid balance inconsistencies in future PRs (who set those so low?)
refactor: optimized fullauto code
fix: stranger gun fire delay fixed (values were swapped in code)
/:cl:

### CODE REVIEW

**Comment by Humonitarian** [comment]
Ready to merge now

**Comment by SirRichardFrancis** [comment]
Fix and balance changes in one package. Classic.

**Comment by Humonitarian** [comment]
> balance

Literally how?

**Comment by SirRichardFrancis** [comment]
> Literally how?

image
image
image
image
image
image

No way that inconsistent and semi-random value changes on selected few guns would change anything...

**Comment by SirRichardFrancis** [comment]
Is balance changes so much of a habit, that you literally don't notice doing it anymore?

**Comment by Humonitarian** [comment]
> > Literally how?
> 
> image image image image image image
> 
> No way those semi-random value tweaks on selected few guns would change anything...

Yeah they change nothing. They were all halved due to the removal of tick_lag from fullauto firedelay, which was a mistake by me.
Haven't you read the PR?

**Comment by SirRichardFrancis** [comment]
FULL_AUTO_300 - fire_delay from 4 to 2, halved
FULL_AUTO_400 - fire_delay from 3 to 1.5, halved
FULL_AUTO_600 - fire_delay from 2 to 1, halved
FULL_AUTO_800 - fire_delay from 1.6 to 0.75, more than halved

BURST_3_ROUND_RAPID - burst_delay from 0.5 to 1.1, delay between shots more than doubled

"SA \"VJP\"" - burst_delay from 0.8 to 1.1 ~33% increase

"unknown plasma gun" - values on single shot and burst switched

"Excelsior HMG .30 \"Maxim\"" - burst_delay from 0.8 to 1.1 ~33% increase
"Excelsior HMG .30 \"Maxim\"" - fire_delay from 1.2 to 2.4, halved

Have no time to check rn, but are those the only guns with fullauto? Would that affect nothing?
Only thing you had to do is replace spawn() with sleep(), that's 100% reasonable. The rest is not so much.

**Comment by ValoTheValo** [comment]
wont this effect tesla chip guns?

**Comment by Humonitarian** [comment]
> wont this effect tesla chip guns? making them dump entire m cells in half a second

No

**Comment by Humonitarian** [comment]
> FULL_AUTO_300 - fire_delay from 4 to 2, halved FULL_AUTO_400 - fire_delay from 3 to 1.5, halved FULL_AUTO_600 - fire_delay from 2 to 1, halved FULL_AUTO_800 - fire_delay from 1.6 to 0.75, more than halved
> 
> BURST_3_ROUND_RAPID - burst_delay from 0.5 to 1.1, delay between shots more than doubled
> 
> "SA "VJP"" - burst_delay from 0.8 to 1.1 ~33% increase
> 
> "unknown plasma gun" - values on single shot and burst switched
> 
> "Excelsior HMG .30 "Maxim"" - burst_delay from 0.8 to 1.1 ~33% increase "Excelsior HMG .30 "Maxim"" - fire_delay from 1.2 to 2.4, halved
> 
> Have no time to check rn, but are those the only guns with fullauto? Would that affect nothing? Only thing you had to do is replace spawn() with sleep(), that's 100% reasonable. The rest is not so much.

Didn't notice I messed up the 800 one, thanks for pointing that out

File: code/__DEFINES/weapons.dm [hunk 1]
```diff
@@ -54,10 +54,10 @@
...
-#define FULL_AUTO_400		list(mode_name = "full auto",  mode_desc = "400 rounds per minute",   mode_type = /datum/firemode/automatic, fire_delay = 3, icon="auto")
-#define FULL_AUTO_600		list(mode_name = "full auto",  mode_desc = "600 rounds per minute",   mode_type = /datum/firemode/automatic, fire_delay = 2  , icon="auto")
-#define FULL_AUTO_800		list(mode_name = "fuller auto",  mode_desc = "800 rounds per minute",   mode_type = /datum/firemode/automatic, fire_delay = 1.6, icon="auto")
+#define FULL_AUTO_300		list(mode_name = "full auto",  mode_desc = "300 rounds per minute",   mode_type = /datum/firemode/automatic, fire_delay = 2  , icon="auto")
+#define FULL_AUTO_400		list(mode_name = "full auto",  mode_desc = "400 rounds per minute",   mode_type = /datum/firemode/automatic, fire_delay = 1.5, icon="auto")
+#define FULL_AUTO_600		list(mode_name = "full auto",  mode_desc = "600 rounds per minute",   mode_type = /datum/firemode/automatic, fire_delay = 1  , icon="auto")
+#define FULL_AUTO_800		list(mode_name = "fuller auto",  mode_desc = "800 rounds per minute",   mode_type = /datum/firemode/automatic, fire_delay = 0.75, icon="auto")
```
Comment (Humonitarian) [inline code/__DEFINES/weapons.dm:60]:
```suggestion
#define FULL_AUTO_800		list(mode_name = "fuller auto",  mode_desc = "800 rounds per minute",   mode_type = /datum/firemode/automatic, fire_delay = 0.8, icon="auto")
```

Comment (Alliostra) [reply code/__DEFINES/weapons.dm]: Humon this is your own PR

Comment (Humonitarian) [reply code/__DEFINES/weapons.dm]: This was easier than editing the file from phone

**Comment by Humonitarian** [comment]
Going to investigate burst delay logic tomorrow to see if that's what kirov means

**Comment by Humonitarian** [comment]
> image To clarify, I requested none of it, that was ~voices in his head~ a bravely defeated strawman.

Guhh?

**Comment by Humonitarian** [comment]
As a conclusion of testing, it seems I was correct. I'll consult firefox for a solution to fix the minimum firerate without turning Eris into a bullet hell, as even with the unkiroved values they would fire too fast.

**Comment by SirRichardFrancis** [comment]
> As a conclusion of testing, it seems I was correct.

At what point??

> even with the unkiroved values they would fire too fast.

I wanted only one thing from you - leaving firerate alone. Repeating this fifth time now - this PR is screwing up recoil in unpredictable ways for no gain. Don't do this.

**Replacing `spawn()` with `sleep()` is all this PR should contain, just that one line. Literally nothing else.**

image
image
image
image

Sixth time, I want you to NOT change any value at all. Only `spawn()` proc with `sleep()` proc. Do you copy?

**Comment by Humonitarian** [comment]
That's the issue. You're of the belief that changing spawn to sleep is the main change in this PR, when it is just one of a few related fixes.

**Comment by Humonitarian** [comment]
A previously existing bug's removal will still cause potentially noticable firerate increase. Nevertheless, we should no longer see firerates of the likes Kirov's video shows. A second atomized PR will come up shortly containing balance changes. How much I hate balance.

**Comment by SirRichardFrancis** [comment]
> That's the issue. You're of the belief that changing spawn to sleep is the main change in this PR, when it is just one of a few related fixes.

Then this PR shouldn't exist, nothing personal, it's just bad.

> How much I hate balance.

Then leave it alone, it needs no changes that screw it up randomly or arbitrary. Just stop doing that.
Eris has morbillion things to do, the ones that would be actually beneficial.

**Comment by Humonitarian** [comment]
> > That's the issue. You're of the belief that changing spawn to sleep is the main change in this PR, when it is just one of a few related fixes.
> 
> Then this PR shouldn't exist, nothing personal, it's just bad.
> 
> > How much I hate balance.
> 
> Then leave it alone, it needs no changes that screw it up randomly or arbitrary. Just stop doing that. Eris has morbillion things to do, the ones that would be actually beneficial.

I'm doing this because of my OCD

**Review by Pink-Chink — APPROVED** [review]
Merged this, then #7878 on test. No obvious issues.

**Comment by Iglpigl** [comment]
> > > That's the issue. You're of the belief that changing spawn to sleep is the main change in this PR, when it is just one of a few related fixes.
> > 
> > 
> > Then this PR shouldn't exist, nothing personal, it's just bad.
> > > How much I hate balance.
> > 
> > 
> > Then leave it alone, it needs no changes that screw it up randomly or arbitrary. Just stop doing that. Eris has morbillion things to do, the ones that would be actually beneficial.
> 
> I'm doing this because of my OCD

https://www.mentalhealth.gov/get-help/immediate-help

**Pull request merged**